Initial documentation on Apparent and True species (#169)
nawi-hub/proteuslib: docs/how_to_guides/how_to_use_apparent_and_true_chemical_species.rst

How to use apparent and true chemical species
=============================================

In ProteusLib, most all chemical processes simulated will be considered "true"

> species, i.e., species that actually exist in an aqueous solution. However, their

Reviewer: ```suggestion
species, i.e., species that actually exist in an aqueous solution (e.g., Na+ and Cl- for NaCl). However, their
```
Already make sense and is clear, but I think it would be nice to spell this out as much as possible.

> species, i.e., species that actually exist in an aqueous solution (e.g., Na+ and Cl- for NaCl). However, their

Reviewer: ```suggestion
species, i.e., species that actually exist in an aqueous solution (e.g., Na+ and Cl- for NaCl). However, there
```
Sorry to drag this out but just noticed this as I was giving one more pass.

may be times when you want to have your system report the "apparent" species
between your inlet and/or outlet ports of a flowsheet. To do this, you need
to understand how to define a species in a given configuration file as "apparent",
which is a special component type in the **thermo-properties** configuration dictionary.

For more information on creating a **thermo-properties** configuration dictionary,
see 'How to setup simple chemistry'.

What you need to update in the thermo-properties configuration dictionary
-------------------------------------------------------------------------

1. Add the ``"Apparent"`` and ``"StateIndex"`` objects under your import statements

> 2. Specify specific species as ``"Apparent"`` within its component definition

Reviewer: ```suggestion
2. Specify certain species as ``"Apparent"`` within its component definition
```

3. Give a list of "true" species it will dissociate into
4. Define a "state_components" argument in the config and specify ``"StateIndex.true"``

Example Configuration Dictionary
--------------------------------

This is best understood by going through an example. In this case, we will consider

> building off of the 'How to setup simple chemistry' guide and add ``"NaCl"`` as

Reviewer: could add a hyperlink to this how-to (even though it's the doc right before this one in the toc tree)

Author: I wanted to do that, but trouble figuring it out. Is there a resource/guide for doing this?

Reviewer: I also had trouble figuring this out. It ends up that doing the following would work:
```suggestion
building off of the :ref:`How to setup simple chemistry<how_to_setup_simple_chemistry>` guide and add ``"NaCl"`` as
```
So the syntax is :ref: `Label title<file_name>` in this case. You can also find more helpful info on Sphinx links [here](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#hyperlinks) and [here](https://sublime-and-sphinx-guide.readthedocs.io/en/latest/references.html).

Author: @adam-a-a It looks like after I made this change the building of the docs failed.

Reviewer: Strange- I tested this locally before suggesting and didn't get any error; maybe I forgot to run `make clean` to start fresh before running `make html` again and missed the warning. Now when I try it locally, I also get the same error.

In any case, adding the following line to the top of the "how to setup simple chemistry" doc should (hopefully) resolve the issue:
`.. _how_to_setup_simple_chemistry:
`

  from idaes.core import AqueousPhase

  # Add Apparent in this import statement
  from idaes.core.components import Solvent, Cation, Anion, Apparent

  import idaes.generic_models.properties.core.pure.Perrys as Perrys
  from idaes.generic_models.properties.core.pure.ConstantProperties import Constant
  from idaes.generic_models.properties.core.state_definitions import FTPx
  from idaes.generic_models.properties.core.eos.ideal import Ideal

  # Add the import for StateIndex
  from idaes.generic_models.properties.core.generic.generic_property import StateIndex

  # Configuration dictionary
  thermo_config = {
      "components": {
          'H2O': {"type": Solvent,
                # Define the methods used to calculate the following properties
                "dens_mol_liq_comp": Perrys,
                "enth_mol_liq_comp": Perrys,
                "cp_mol_liq_comp": Perrys,
                "entr_mol_liq_comp": Perrys,
                # Parameter data is always associated with the methods defined above
                "parameter_data": {
                      "mw": (18.0153, pyunits.g/pyunits.mol),
                      # Parameters here come from Perry's Handbook:  p. 2-98

>                       "dens_mol_liq_comp_coeff": {

Reviewer: Now that I think about it, is there a way for the user to avoid having to specify these parameter data from Perry's Handbook in this example? Is that where the EDB will come into play? I was just thinking that if a user didn't have these parameter values readily available, it could be painful... but I am assuming that the database (or some other means) would be used to avoid manually entering such values (at least for common configurations).

Reviewer: On the other hand, this is useful because it shows the user how to manually enter such parameters.

Author: @adam-a-a  Yes, eventually much of this will be done by EDB. EDB is in parallel development with the rest of the chem team, so for the time being I am doing this stuff manually. Future documentation for how to do this with EDB is going to come (#140).

Reviewer 2: We need to be careful with this however - we need to make sure we have the right to reproduce data like this. Also, the IDAES approach has generally be to provide a way for users to input data, but to leave finding the data to the user.

Author: Sounds to me that, at least for the purpose of this example, we should just use any open source parameters or methods. The exact values of parameters are not important for any of these docs.

                                    },
                            # End parameter_data
                            },
                },
                # End Component list

          "phases":  {'Liq': {"type": AqueousPhase,
                              "equation_of_state": Ideal},
                      },

          "state_definition": FTPx,
          "state_bounds": {"flow_mol": (0, 50, 100),
                           "temperature": (273.15, 300, 650),
                           "pressure": (5e4, 1e5, 1e6)
                       },

          # We must define the 'StateIndex' as "true". This is because in ProteusLib,
          #   all speciation reactions are defined on the true species, not the
          #   apparent species.
          "state_components": StateIndex.true,

          "pressure_ref": 1e5,
          "temperature_ref": 300,
          "base_units": {"time": pyunits.s,
                         "length": pyunits.m,
                         "mass": pyunits.kg,
                         "amount": pyunits.mol,
                         "temperature": pyunits.K},
      }
      # End thermo_config definition

.. note::

>     When you define a species as ``"Apparent"``, and specify ``"state_components": StateIndex.true``,

Reviewer: ```suggestion
    When you define a species as ``"Apparent"`` and specify ``"state_components": StateIndex.true``,
```
